clients/java/signalr/src/main/java/HubConnection.java

```diff
@@ -23,19 +23,32 @@ public HubConnection(String url, Transport transport){
         this.protocol = new JsonHubProtocol();
         this.callback = (payload) -> {
 
-            InvocationMessage[] messages = protocol.parseMessages(payload);
-
-            // message will be null if we receive any message other than an invocation.
-            // Adding this to avoid getting error messages on pings for now.
-            for (InvocationMessage message : messages) {
-                if (message != null && handlers.containsKey(message.target)) {
-                    ArrayList<Object> args = gson.fromJson((JsonArray)message.arguments[0], (new ArrayList<Object>()).getClass());
-                    List<ActionBase> actions = handlers.get(message.target);
-                    if (actions != null) {
-                        for (ActionBase action: actions) {
-                            action.invoke(args.toArray());
+            HubMessage[] messages = protocol.parseMessages(payload);
+
+            for (HubMessage message : messages) {
+                switch (message.getMessageType()) {
```

Comment:
There may be a performance or design reason behind this, but why have this enum when you can just call a generic method on each message type, like:
```java
for (HubMessage message : messages) {
    message.Handle(params);
}
```

And then in the class itself invoke the actions.

Reply:
I see that this is how it was done in the C# client, so you can probably dismiss this comment.